### PR TITLE
[afreecatv] Add support for livestreams

### PR DESCRIPTION
**Pull request by wlritchi**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

AfreecaTV is a platform for both prerecorded and live video content. Until now, `yt-dlp` has only supported downloading VODs from AfreecaTV. This PR adds a new extractor for live video from AfreecaTV. The new extractor inherits from the existing AfreecaTV extractor in order to reuse its authentication logic.

### CODE REVIEW

File: supportedsites.md [hunk 1]
```diff
@@ -40,6 +40,7 @@
  - **aenetworks:collection**
  - **aenetworks:show**
  - **afreecatv**: afreecatv.com
+ - **afreecatv:live**: afreecatv.com
```
Comment (Ashish0804) [inline supportedsites.md:43]: this is done automatically....u can remove this
```suggestion
```

File: yt_dlp/extractor/afreecatv.py [hunk 2]
```diff
@@ -380,3 +383,177 @@ def _real_extract(self, url):
...
+            },
+            data=urlencode_postdata(form_data),
+        )
+        channel_info = info['CHANNEL']
+        broadcast_no = channel_info.get('BNO') or broadcast_no
+        form_data['bno'] = broadcast_no
+        form_data['type'] = 'aid'
+
+        formats = []
+        # TODO do we need to consider other stream key types for different formats?
```
Comment (Ashish0804) [inline yt_dlp/extractor/afreecatv.py:479]: if this is not done then mark this PR as draft, else remove this

File: yt_dlp/extractor/afreecatv.py [hunk 3]
```diff
@@ -380,3 +383,177 @@ def _real_extract(self, url):
...
+        'original',
+    ]
+
+    def _real_extract(self, url):
+        match = self._match_valid_url(url)
+        broadcaster_id = match.group('id')
+        broadcast_no = match.group('bno')
+
+        # note: the only necessary input is "bid" in the form data
+        # the rest, including the query param, is just to keep parity with the website
```
Comment (Ashish0804) [inline yt_dlp/extractor/afreecatv.py:454]: not needed

File: yt_dlp/extractor/afreecatv.py [hunk 4]
```diff
@@ -380,3 +383,177 @@ def _real_extract(self, url):
...
+        form_data = {
+            'bid': broadcaster_id,
+            'type': 'live',
+            'player_type': 'html5',
+            'stream_type': 'common',
+            'mode': 'landing',
+            'from_api': 0,
```
Comment (Ashish0804) [inline yt_dlp/extractor/afreecatv.py:461]: try to minimize the required fields

File: yt_dlp/extractor/afreecatv.py [hunk 5]
```diff
@@ -380,3 +383,177 @@ def _real_extract(self, url):
...
+                stream_info = self._download_json(
+                    stream_info_url,
+                    broadcast_no,
+                    query={
+                        'return_type': cdn,
+                        'use_cors': True,
+                        'cors_origin_url': 'play.afreecatv.com',
+                        'broad_key': stream_key,
+                        # TODO the website passes a "time" parameter, but it's not obvious how it's derived
+                        # for now we can get away with ignoring it
```
Comment (Ashish0804) [inline yt_dlp/extractor/afreecatv.py:509]: `time.time()*1000` maybe 🤔

File: yt_dlp/extractor/afreecatv.py [hunk 6]
```diff
@@ -380,3 +383,177 @@ def _real_extract(self, url):
...
+            )
+            info['upload_date'] = unified_strdate(station_info.get('broad_start'))
+
+            if not info['title']:
+                # not exactly equivalent to stream title, use only as fallback
+                info['title'] = station_info.get('station_title')
+            if not info['uploader']:
+                info['uploader'] = station_info.get('station_name')
+        except Exception as e:
+            self.report_warning('Failed to look up extra metadata: ' + error_to_compat_str(e))
```
Comment (Ashish0804) [inline yt_dlp/extractor/afreecatv.py:557]: no need for this to be in try-except block.
Just let it be `None` if extraction failed

Comment (Ashish0804) [reply yt_dlp/extractor/afreecatv.py:557]: Remove all these and then inline the info dict

File: yt_dlp/extractor/afreecatv.py (hunk 4 above)
Comment (Ashish0804) [reply yt_dlp/extractor/afreecatv.py:439]: No.
Use the bare minimum.

File: yt_dlp/extractor/afreecatv.py [hunk 7]
```diff
@@ -380,3 +382,147 @@ def _real_extract(self, url):
...
+                    'format_id': quality_str,
+                    'quality': quality_key(quality_str),
+                    'ext': 'mp4',
+                    'protocol': 'm3u8',
+                    'url': update_url_query(view_url, {
+                        'aid': aid,
+                    }),
+                })
+            except (ExtractorError, KeyError):
+                continue
```
Comment (Ashish0804) [inline yt_dlp/extractor/afreecatv.py:506]: remove try-except here aswell.

set `fatal=False` for api requests and then use `.get` or `try_get` to get the fields.
As of making sure shit is not appended to formats, make a check for all the necessary keys and only append if they are not None

File: yt_dlp/extractor/afreecatv.py [hunk 8]
```diff
@@ -380,3 +384,90 @@ def _real_extract(self, url):
             })
 
         return info
+
+
+class AfreecaTVLiveIE(AfreecaTVIE):
+
+    IE_NAME = 'afreecatv:live'
+    _VALID_URL = r'https?://play\.afreeca(?:tv)?\.com(?::\d+)?/(?P<id>[^/]+)(?:/(?P<bno>\d+))?'
```
Comment (pukkandan) [inline yt_dlp/extractor/afreecatv.py:392]: Add tests for the other URL patterns. `only_matching` tests are sufficient

**Review by pukkandan — CHANGES_REQUESTED** [review]
Also, please make sure I didn't break anything

**Comment by wlritchi** [comment]
Everything looks to be working after the cleanup :)

Optimization question: The extractor needs to make a couple of API calls per format in order to get the M3U8 URLs for each. Is there a way to do these calls lazily, or some other optimization the extractor can apply, so that e.g. with `-f best` we only fetch the URL for the highest quality?

**Comment by pukkandan** [comment]
> The extractor needs to make a couple of API calls per format in order to get the M3U8 URLs for each. Is there a way to do these calls lazily,

No

> or some other optimization the extractor can apply, so that e.g. with -f best we only fetch the URL for the highest quality?

You can create an `--extractor-args` to let the user specify which formats to fetch

**Comment by pukkandan** [comment]
I am merging this. You can add extractor args in another PR if needed

**Comment by wlritchi** [comment]
:+1: I think in this case it's probably not worth it; it's a minor optimization that probably isn't really worth my time learning about the custom extractor args. Maybe I'll come back to it in the future if a refactor adds a common way to do it. :)

**Pull request merged**